server/lib/timeline.ts

```diff
@@ -0,0 +1,200 @@
+          ActivityTypes.COLLECTIVE_VIRTUAL_CARD_SUSPENDED_DUE_TO_INACTIVITY,
+          ActivityTypes.VIRTUAL_CARD_PURCHASE,
+        ],
+        UserId: user.id,
+      });
+    }
+    if (classes.includes(ActivityClasses.CONTRIBUTIONS)) {
+      conditionals.push({
+        type: [
+          ActivityTypes.PAYMENT_FAILED,
+          ActivityTypes.ORDER_PAYMENT_FAILED,
+          ActivityTypes.ORDER_THANKYOU,
+          ActivityTypes.ORDER_PROCESSING,
+        ],
+        [Op.or]: [{ UserId: collective.CreatedByUserId }, { FromCollectiveId: collective.id }],
+      });
+    }
+    if (classes.includes(ActivityClasses.ACTIVITIES_UPDATES)) {
+      conditionals.push({
+        type: ActivityTypes.COLLECTIVE_UPDATE_PUBLISHED,
+        CollectiveId: getCollectiveIdsForRole(memberships, [
+          MemberRoles.BACKER,
+          MemberRoles.FOLLOWER,
+          MemberRoles.MEMBER,
+          MemberRoles.CONTRIBUTOR,
+          MemberRoles.ATTENDEE,
+          MemberRoles.ADMIN,
+        ]),
+      });
+    }
+    return { [Op.or]: conditionals };
+  } else {
+    return { [Op.or]: [{ CollectiveId: collective.id }, { FromCollectiveId: collective.id }] };
+  }
+};
+
+const order: Order = [['createdAt', 'DESC']];
+const FEED_LIMIT = 1000;
+
+const generateFeed = async (collective: Collective, sinceId?: string) => {
+  const redis = await createRedisClient();
+  const cacheKey = `timeline-${collective.slug}`;
+
+  const where = await generateTimelineQuery(collective);
+  if (sinceId) {
+    where['id'] = { [Op.gt]: sinceId };
+  }
+  const result = await models.Activity.findAll({ where, order, limit: FEED_LIMIT });
+  if (result.length > 0) {
+    const activities = result.map(({ id, type }) => ({ score: id, value: JSON.stringify({ id, type }) }));
+    debug(`Generated timeline for ${collective.slug} with ${activities.length} activities`);
+    await redis.zAdd(cacheKey, activities);
+
+    // Trim the cache if updating with new activities
+    if (sinceId) {
+      const count = await redis.zCount(cacheKey, '0', '+inf');
+      if (count > FEED_LIMIT) {
+        await redis.zRemRangeByRank(cacheKey, 0, count - FEED_LIMIT - 1);
+      }
+    }
+  }
+};
+
+export const getCollectiveFeed = async ({
+  collective,
+  untilId,
+  limit,
+  classes,
+}: {
+  collective: Collective;
+  untilId: number;
+  limit: number;
+  classes: ActivityClasses[];
+}) => {
+  const redis = await createRedisClient();
+  // If we don't have a redis client, we can't cache the timeline using sorted sets
+  if (!redis) {
+    debug('Redis is not configured, skipping cached timeline');
+    const where = await generateTimelineQuery(collective, classes);
+    if (untilId) {
+      where['id'] = { [Op.lt]: untilId };
+    }
+    return models.Activity.findAll({
+      where,
+      order,
+      limit,
+    });
+  }
+
+  // Check if timeline cache exists
+  const cacheKey = `timeline-${collective.slug}`;
+  const cacheExists = await redis.exists(cacheKey);
+  if (!cacheExists) {
```

Comment:
I'm wondering if we still want to call this a `cache`.

Reply:
If we were to consider the data structure used it would be a sorted set but it is still cached data. I think `cache` conveys better what is going on here.